internal/queue/queue.go

```diff
@@ -0,0 +1,75 @@
+package queue
+
+// Queue[T] is a segmented FIFO queue of Ts.
+type Queue[T any] struct {
+	next  *Queue[T]
+	items []*T
+	head  int
+	tail  int
+}
+
+// New creates a new instance of Queue[T].
+//   - size is the size of each Queue segment
+func New[T any](size int) *Queue[T] {
+	return &Queue[T]{
+		items: make([]*T, size),
+	}
+}
+
+// Enqueue adds the specified item to the end of the queue.
+// If the current segment is full, a new segment is created.
+func (q *Queue[T]) Enqueue(item T) {
+	cur := q
+	for {
+		// always enqueue to the last segment
```

Comment:
This is a bit broken.  If there are multiple segments and they are all empty, we should enqueue to the first segment.

Reply:
Will also fix the operations to be O(1).

Reply:
Fixed.